### ISSUE TITLE
`explain()` crashes when using combined approaches on categorical and mixed data.

### ISSUE THREAD

**Issue by LHBO** [NorskRegnesentral/shapr#362]
See code below which illustrates that the `explain()` function crashes sometimes. It seems to depend on the order of the approaches. The code still crashes after applying the code from https://github.com/NorskRegnesentral/shapr/pull/356.

The error message we get is 
```
Error in `[.data.table`(dt, , lapply(.SD, function(x) sum(((x) * w)/sum(w))),  : 
  Some items of .SDcols are not column names: [p_hat1]
In addition: Warning message:
In `[.data.table`(dt, , `:=`((pred_cols), predict_model(model, newdata = .SD)),  :
  Invalid .internal.selfref detected and fixed by taking a (shallow) copy of the data.table so that := can add this new column by reference. At an earlier point, this data.table has been copied by R (or was created manually using structure() or similar). Avoid names<- and attr<- which in R currently (and oddly) may copy the whole data.table. Use set* syntax instead to avoid copying: ?set, ?setnames and ?setattr. If this message doesn't help, please report your use case to the data.table issue tracker so the root cause can be fixed or this message improved.
```

The error occurs from the `compute_MCint()` function in `compute_vS.R`. 
But this is due to that in `batch_compute_vS()`, when calling `compute_MCint(dt, pred_cols)` , the `dt` data.table object does not always contain the `pred_cols` columns.
This is because the `compute_preds()` function tries to update the `dt` object by reference but is not able to do this due to Invalid .internal.selfref  (see warning above).  
A simple fix is to use the returned object from `compute_preds()` instead of updating by reference. Then only the warning will still occur. This can be fixed by adding `dt = data.table::copy(dt)` inside `compute_preds()` to make a local copy inside the function.
DISCUSS WITH MARTIN WHAT HE THINKS.
I can make a PR with the proposed fixes above.

An example which causes `explain()` to crash.

```
options(digits = 5) # To avoid round off errors when printing output on different systems

set.seed(12345)
library(data.table)

data("airquality")
data <- data.table::as.data.table(airquality)
data[, Month_factor := as.factor(Month)]
data[, Ozone_sub30 := (Ozone < 30) * 1]
data[, Ozone_sub30_factor := as.factor(Ozone_sub30)]
data[, Solar.R_factor := as.factor(cut(Solar.R, 10))]
data[, Wind_factor := as.factor(round(Wind))]

data_complete <- data[complete.cases(airquality), ]
data_complete <- data_complete[sample(seq_len(.N))] # Sh

y_var_numeric <- "Ozone"
y_var_binary <- "Ozone_sub30"
y_var_binaryfactor <- "Ozone_sub30_factor"

x_var_numeric <- c("Solar.R", "Wind", "Temp", "Month", "Day")
x_var_mixed <- c("Solar.R", "Wind", "Temp", "Day", "Month_factor")
x_var_categorical <- c("Month_factor", "Ozone_sub30_factor", "Solar.R_factor", "Wind_factor")

data_train <- head(data_complete, -3)
data_explain <- tail(data_complete, 3)

x_train_numeric <- data_train[, ..x_var_numeric]
x_train_mixed <- data_train[, ..x_var_mixed]
x_train_categorical <- data_train[, ..x_var_categorical]

x_explain_numeric <- data_explain[, ..x_var_numeric]
x_explain_mixed <- data_explain[, ..x_var_mixed]
x_explain_categorical <- data_explain[, ..x_var_categorical]

lm_formula_numeric <- as.formula(paste0(y_var_numeric, " ~ ", paste0(x_var_numeric, collapse = " + ")))
lm_formula_mixed <- as.formula(paste0(y_var_numeric, " ~ ", paste0(x_var_mixed, collapse = " + ")))
lm_formula_interaction <- Ozone ~ Solar.R * Wind
# lm_formula_numeric_col_order <- as.formula(paste0(y_var_numeric, " ~ ",
#                                                   paste0(sort(x_var_numeric), collapse = " + ")))
lm_formula_categorical <- as.formula(paste0(y_var_numeric, " ~ ", paste0(x_var_categorical, collapse = " + ")))

model_lm_numeric <- lm(lm_formula_numeric, data = data_complete)
model_lm_categorical <- lm(lm_formula_categorical, data = data_complete)
model_lm_mixed <- lm(lm_formula_mixed, data = data_complete)

p0 <- data_train[, mean(get(y_var_numeric))]


# The "works" and "do not work" is based on the master at the date the issue was created.
# They are not necessarily correct for newer versions 

# works
explain_numeric <- explain(
  model = model_lm_numeric,
  x_explain = x_explain_numeric,
  x_train = x_train_numeric,
  approach = c("independence", "ctree", "ctree", "independence", "independence"),
  prediction_zero = p0,
  n_batches = NULL,
  timing = FALSE
)

# works
explain_mixed <- explain(
  model = model_lm_mixed,
  x_explain = x_explain_mixed,
  x_train = x_train_mixed,
  approach = c("ctree", "independence", "independence", "independence", "independence"),
  prediction_zero = p0,
  n_batches = NULL,
  timing = FALSE
)

# do not work
explain_mixed <- explain(
  model = model_lm_mixed,
  x_explain = x_explain_mixed,
  x_train = x_train_mixed,
  approach = c("independence", "ctree", "ctree", "ctree", "ctree"),
  prediction_zero = p0,
  n_batches = NULL,
  timing = FALSE
)

# works
explain_categorical <- explain(
  model = model_lm_categorical,
  x_explain = x_explain_categorical,
  x_train = x_train_categorical,
  approach = c("ctree", "independence", "independence", "independence"),
  prediction_zero = p0,
  n_batches = NULL,
  timing = FALSE
)

# works
explain_categorical <- explain(
  model = model_lm_categorical,
  x_explain = x_explain_categorical,
  x_train = x_train_categorical,
  approach = c("ctree", "ctree", "independence", "independence"),
  prediction_zero = p0,
  n_batches = NULL,
  timing = FALSE
)

# works
explain_categorical <- explain(
  model = model_lm_categorical,
  x_explain = x_explain_categorical,
  x_train = x_train_categorical,
  approach = c("ctree", "independence", "ctree", "independence"),
  prediction_zero = p0,
  n_batches = NULL,
  timing = FALSE
)

# works
explain_categorical <- explain(
  model = model_lm_categorical,
  x_explain = x_explain_categorical,
  x_train = x_train_categorical,
  approach = c("ctree", "ctree", "ctree", "independence"),
  prediction_zero = p0,
  n_batches = NULL,
  timing = FALSE
)

# do not work
explain_categorical <- explain(
  model = model_lm_categorical,
  x_explain = x_explain_categorical,
  x_train = x_train_categorical,
  approach = c("independence", "ctree", "ctree", "ctree"),
  prediction_zero = p0,
  n_batches = NULL,
  timing = FALSE
)

# do not work
explain_categorical <- explain(
  model = model_lm_categorical,
  x_explain = x_explain_categorical,
  x_train = x_train_categorical,
  approach = c("independence", "ctree", "independence", "ctree"),
  prediction_zero = p0,
  n_batches = NULL,
  timing = FALSE
)

# do not work
explain_mixed <- explain(
  model = model_lm_mixed,
  x_explain = x_explain_mixed,
  x_train = x_train_mixed,
  approach = c("independence", "ctree", "independence", "independence", "ctree"),
  prediction_zero = p0,
  n_batches = NULL,
  timing = FALSE
)

```

**Comment by LHBO** [comment]
A better solution is to figure out why the updating by reference does not always work.

**Comment by LHBO** [comment]
This problem is fixed in #356. 
I will close this issue for now.
